File: typescript/ch7/memoize.ts
import fp from 'lodash/fp';
import _ from 'lodash';
import checkPerformance from './checkPerformance';

function memoize() {
  // FP 에서는 이런 식으로 외부 변수가 있으면 안 되지만 테스트 결과를 위해서 임시로 사용한다..
  let count = 0;
  const add = (a: number, b: number) => {
    count += 1;
    return a + b;
  }
  const addMemoized = fp.memoize(add);

  const logCount = () => console.log(count);
  const resetCount = () => { count = 0; }
  const loop = (func: any) => _.range(10000000).forEach(func);
  const checkPerformancePlusCount = (name: string, func: any) => fp.compose(
    logCount,
    () => checkPerformance(name, func),
    resetCount
  )();

  checkPerformancePlusCount('add', () => loop(() => add(1, 2)));
  checkPerformancePlusCount('addMemoized', () => loop(() => addMemoized(1, 2)));
  
}

export {
  memoize
};

File: python/python/ch7/memoize.py
import pydash
import toolz as tz
from .check_performance import check_performance

global count
count = 0

def memoize():
    def add(a, b):
        global count
        count += 1
        return a + b
      
    def log_count(): 
        global count
        print(count)
        
    def reset_count():
        global count
        count = 0
        
    def loop(func):
        pydash.for_each(range(1000000), func)
    
    add_memoized = pydash.memoize(add)
    
    def check_performance_plus_count(name, func):
        return tz.compose(
            lambda v: log_count(),
            lambda v: check_performance(name, func),
            reset_count
        )()
        
    check_performance_plus_count('add', lambda: loop(lambda: add(1, 2)))
    check_performance_plus_count('add_memoized', lambda: loop(lambda: add_memoized(1, 2)))